File: GuardianMonsters/core/src/de/limbusdev/guardianmonsters/CoreServiceLocator.java
package de.limbusdev.guardianmonsters;

import de.limbusdev.guardianmonsters.services.IGameStateService;
import de.limbusdev.guardianmonsters.services.NullGameStateService;

public class CoreServiceLocator
{
    public interface Service
    {
        /**
         * If using Singletons as service, set instance null on destroy
         */
        void destroy();
    }

    private static IGameStateService gameState;

    public static void provide(IGameStateService service) {
        gameState = service;
    }

    public static IGameStateService getGameState() {
        if(gameState == null) {
            System.err.println("SERVICES: No Game State service injected yet with Services.provide(IGameStateService). Returning NullGameStateService.");
            return new NullGameStateService();
        } else {
            return gameState;
        }
    }


    public static void destroy()
    {
        if(gameState != null) {
            gameState.destroy();
            gameState = null;
        }
    }
}

File: GuardianMonsters/core/src/de/limbusdev/guardianmonsters/CoreServiceLocator.kt
package de.limbusdev.guardianmonsters

import de.limbusdev.guardianmonsters.services.IGameStateService
import de.limbusdev.guardianmonsters.services.NullGameStateService

object CoreServiceLocator {

    private var gameState: IGameStateService? = null

    interface Service
    {
        /** If using Singletons as service, set instance null on destroy */
        fun destroy()
    }

    fun provide(service: IGameStateService) { gameState = service }

    fun getGameState(): IGameStateService
    {
        return if (gameState == null)
        {
            System.err.println("SERVICES: No Game State service injected yet with Services.provide(IGameStateService). Returning NullGameStateService.")
            NullGameStateService()
        }
        else
        { gameState as IGameStateService }
    }


    fun destroy()
    {
        if (gameState != null)
        {
            gameState!!.destroy()
            gameState = null
        }
    }
}